velox/connectors/hive/PartitionIdGenerator.cpp

```diff
@@ -0,0 +1,139 @@
+
+namespace facebook::velox::connector::hive {
+
+PartitionIdGenerator::PartitionIdGenerator(
+    const RowTypePtr& inputType,
+    std::vector<column_index_t> partitionChannels,
+    memory::MemoryPool* FOLLY_NONNULL pool)
+    : partitionChannels_(std::move(partitionChannels)),
+      allPartitionsVector_(std::static_pointer_cast<RowVector>(
+          BaseVector::create(inputType, 0, pool))),
+      activeIds_(SelectivityVector::empty(0)),
+      maxId_(0) {
+  hashers_.reserve(partitionChannels_.size());
+  for (column_index_t i = 0; i < partitionChannels_.size(); i++) {
+    hashers_.push_back(exec::VectorHasher::create(
+        inputType->childAt(partitionChannels_[i]), partitionChannels_[i]));
+  }
+}
+
+std::map<uint64_t, uint64_t> PartitionIdGenerator::run(
+    const RowVectorPtr& input,
+    raw_vector<uint64_t>& result) {
+  result.resize(input->size());
+  bool rehashed =
+      computeIds(input, SelectivityVector{input->size(), true}, result);
+
+  std::map<uint64_t, uint64_t> idMap;
+  if (rehashed && allPartitionsVector_->size() > 0) {
+    raw_vector<uint64_t> newIds(allPartitionsVector_->size());
+
+    VELOX_CHECK(
+        !computeIds(allPartitionsVector_, activeIds_, newIds),
+        "Partition ID generator should not rehash when computing for the existing partitions.");
+
+    activeIds_.applyToSelected([&](auto id) { idMap[id] = newIds[id]; });
+
+    movePartitions(idMap);
+  }
+
+  addPartitions(input, result);
+
+  return idMap;
+}
+
+bool PartitionIdGenerator::computeIds(
+    const RowVectorPtr& input,
+    const SelectivityVector& rows,
+    raw_vector<uint64_t>& result) {
+  result.resize(input->size());
+
+  for (auto& hasher : hashers_) {
+    auto partitionVector = input->childAt(hasher->channel())->loadedVector();
+    hasher->decode(*partitionVector, rows);
+  }
+
+  bool rehashed = false, toRehash = false;
+  do {
+    toRehash = false;
+    for (auto& hasher : hashers_) {
+      if (!hasher->computeValueIds(rows, result)) {
+        toRehash = true;
+      }
+    }
+    if (toRehash) {
+      rehashed = true;
+
+      uint64_t multiplier = 1;
+      for (auto i = 0; i < hashers_.size(); i++) {
+        multiplier = hashers_[i]->enableValueIds(multiplier, kHasherReservePct);
+        VELOX_CHECK_NE(
+            multiplier,
+            exec::VectorHasher::kRangeTooLarge,
+            "Exceeded limit of distinct partitions.");
+      }
+    }
+  } while (toRehash);
+
+  return rehashed;
+}
+
+void PartitionIdGenerator::movePartitions(
+    const std::map<uint64_t, uint64_t>& indexMap) {
+  vector_size_t maxNewId = indexMap.rbegin()->second;
+
+  allPartitionsVector_->resize(maxNewId + 1, false);
+  activeIds_.resize(maxNewId + 1, false);
+
+  for (auto entry = indexMap.rbegin(); entry != indexMap.rend(); entry++) {
+    allPartitionsVector_->copy(
+        allPartitionsVector_.get(), entry->second, entry->first, 1);
+    allPartitionsVector_->setNull(entry->first, true);
```

Comment:
Can it happen that old ID and new ID are the same?

Reply:
You're right, should add condition to skip the case of same old ID and new ID. Somehow this part is old version.

Reply:
It seems we might be missing some test coverage. Would it be possible to add more tests to ensure this case is tested?